From a649110b8dce85d06a4f6560856f103545bc94ed Mon Sep 17 00:00:00 2001
From: Maulik K Patel <maulik.k.patel@xilinx.com>
Date: Tue, 27 Aug 2019 06:38:40 -0700
Subject: [PATCH 0667/1851] v4l: xilinx: scd: Fixed resolution switch issue

commit a71da1cee027d94b53c9e7d6ca2e7e6fede95bec from
https://github.com/Xilinx/linux-xlnx.git

The issue is when we switch the resolution of video stream in stream
based SCD mode, SCD IP was not working after changing resolution. So
to fix this issue, SCD IP needs to be reset when closing video stream.

Signed-off-by: Maulik K Patel <maulik.k.patel@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 .../platform/xilinx/xilinx-scenechange-channel.c     | 12 ++++++++++++
 drivers/media/platform/xilinx/xilinx-scenechange.c   |  3 ---
 drivers/media/platform/xilinx/xilinx-scenechange.h   |  3 +++
 3 files changed, 15 insertions(+), 3 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-scenechange-channel.c b/drivers/media/platform/xilinx/xilinx-scenechange-channel.c
index b0270ba848b4..4f3a8d03d217 100644
--- a/drivers/media/platform/xilinx/xilinx-scenechange-channel.c
+++ b/drivers/media/platform/xilinx/xilinx-scenechange-channel.c
@@ -8,6 +8,7 @@
  *          Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
  */
 
+#include <linux/gpio/consumer.h>
 #include <linux/of.h>
 #include <linux/xilinx-v4l2-events.h>
 
@@ -196,11 +197,22 @@ static int xscd_s_ctrl(struct v4l2_ctrl *ctrl)
 static int xscd_s_stream(struct v4l2_subdev *subdev, int enable)
 {
 	struct xscd_chan *chan = to_xscd_chan(subdev);
+	struct xscd_device *xscd = chan->xscd;
 
 	if (enable)
 		xscd_chan_configure_params(chan);
 
 	xscd_dma_enable_channel(&chan->dmachan, enable);
+
+	/*
+	 * Resolution change doesn't work in stream based mode unless
+	 * the device is reset.
+	 */
+	if (!enable && !xscd->memory_based) {
+		gpiod_set_value_cansleep(xscd->rst_gpio, XSCD_RESET_ASSERT);
+		gpiod_set_value_cansleep(xscd->rst_gpio, XSCD_RESET_DEASSERT);
+	}
+
 	return 0;
 }
 
diff --git a/drivers/media/platform/xilinx/xilinx-scenechange.c b/drivers/media/platform/xilinx/xilinx-scenechange.c
index 5d92b24b3eae..38db9c7d37b2 100644
--- a/drivers/media/platform/xilinx/xilinx-scenechange.c
+++ b/drivers/media/platform/xilinx/xilinx-scenechange.c
@@ -17,9 +17,6 @@
 
 #include "xilinx-scenechange.h"
 
-#define XSCD_RESET_DEASSERT	(0)
-#define XSCD_RESET_ASSERT	(1)
-
 static irqreturn_t xscd_irq_handler(int irq, void *data)
 {
 	struct xscd_device *xscd = (struct xscd_device *)data;
diff --git a/drivers/media/platform/xilinx/xilinx-scenechange.h b/drivers/media/platform/xilinx/xilinx-scenechange.h
index d5dfaa02b5e0..5cc9ce54bfc4 100644
--- a/drivers/media/platform/xilinx/xilinx-scenechange.h
+++ b/drivers/media/platform/xilinx/xilinx-scenechange.h
@@ -63,6 +63,9 @@ struct gpio_desc;
 
 #define XSCD_MAX_CHANNELS		8
 
+#define XSCD_RESET_DEASSERT		(0)
+#define XSCD_RESET_ASSERT		(1)
+
 /****************************** PROTOTYPES ************************************/
 
 struct xscd_device;
-- 
2.31.1

